an error of generic sumtype

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:**  latest V 0.2.4 dee4ffb
**OS:**  macos

**What did you do?**
here is the example code:
```
struct Foo<T> {
	x T
}

struct Bar<T> {
	x T
}

type MyType<T> = Bar<T> | Foo<T>


fn main() {
	f := Foo<string>{'hi'}
	t := MyType<string>(f)
	println(t.type_name()) 
	// println(t.x)
}
```

**What did you expect to see?**
no error
 
**What did you see instead?**
report an error:
```
==================
                ^        ~~~~~~~~
/tmp/v/main.9593052193497396179.tmp.c:911:32: error: expected ';' at end of declaration list
                main__Bar<string>* _main__Bar<string>;
                                             ^
                                             ;
2 errors generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/ast/table.v
+++ b/vlib/v/ast/table.v
@@ -1771,10 +1771,18 @@ pub fn (mut t Table) generic_insts_to_concrete() {
 							}
 						}
 						for i in 0 .. variants.len {
-							if t_typ := t.resolve_generic_to_concrete(variants[i], generic_names,
-								info.concrete_types)
-							{
-								variants[i] = t_typ
+							if variants[i].has_flag(.generic) {
+								sym := t.get_type_symbol(variants[i])
+								if sym.kind == .struct_ && variants[i].idx() != info.parent_idx {
+									variants[i] = t.unwrap_generic_type(variants[i], generic_names,
+										info.concrete_types)
+								} else {
+									if t_typ := t.resolve_generic_to_concrete(variants[i],
+										generic_names, info.concrete_types)
+									{
+										variants[i] = t_typ
+									}
+								}
 							}
 						}
 						typ.info = SumType{

--- a/vlib/v/tests/generic_sumtype_insts_test.v
+++ b/vlib/v/tests/generic_sumtype_insts_test.v
@@ -1,0 +1,16 @@
+struct Foo<T> {
+	x T
+}
+
+struct Bar<T> {
+	x T
+}
+
+type MyType<T> = Bar<T> | Foo<T>
+
+fn test_generic_sumtype_insts() {
+	f := Foo<string>{'hi'}
+	t := MyType<string>(f)
+	println(t.type_name())
+	assert t.type_name() == 'Foo<string>'
+}

